File: scitt-golang/pkg/merkle/checkpoints.go
package merkle

import (
	"crypto/ecdsa"
	"encoding/base64"
	"encoding/binary"
	"fmt"
	"net/url"
	"regexp"
	"strings"
	"time"

	"github.com/tradeverifyd/transparency-service/scitt-golang/pkg/cose"
)

// Checkpoint represents a signed tree head
// It contains a commitment to the current state of the Merkle tree
type Checkpoint struct {
	TreeSize  int64         // Number of entries in the tree
	RootHash  [HashSize]byte // Root hash of the Merkle tree
	Timestamp int64         // Unix timestamp in milliseconds
	Origin    string        // Transparency service URL
	Signature []byte        // ES256 signature
}

// CreateCheckpoint creates a signed checkpoint for the current tree state
func CreateCheckpoint(treeSize int64, rootHash [HashSize]byte, privateKey *ecdsa.PrivateKey, origin string) (*Checkpoint, error) {
	// Validate origin URL
	if _, err := url.Parse(origin); err != nil {
		return nil, fmt.Errorf("invalid origin URL: %w", err)
	}

	timestamp := time.Now().UnixMilli()

	// Create the data to be signed
	dataToSign := encodeCheckpointData(treeSize, rootHash, timestamp, origin)

	// Sign with ES256
	signer, err := cose.NewES256Signer(privateKey)
	if err != nil {
		return nil, fmt.Errorf("failed to create signer: %w", err)
	}

	signature, err := signer.Sign(dataToSign)
	if err != nil {
		return nil, fmt.Errorf("failed to sign checkpoint: %w", err)
	}

	return &Checkpoint{
		TreeSize:  treeSize,
		RootHash:  rootHash,
		Timestamp: timestamp,
		Origin:    origin,
		Signature: signature,
	}, nil
}

// VerifyCheckpoint verifies a checkpoint signature
func VerifyCheckpoint(checkpoint *Checkpoint, publicKey *ecdsa.PublicKey) (bool, error) {
	// Reconstruct the signed data
	dataToSign := encodeCheckpointData(
		checkpoint.TreeSize,
		checkpoint.RootHash,
		checkpoint.Timestamp,
		checkpoint.Origin,
	)

	// Verify signature
	verifier, err := cose.NewES256Verifier(publicKey)
	if err != nil {
		return false, fmt.Errorf("failed to create verifier: %w", err)
	}

	valid, err := verifier.Verify(dataToSign, checkpoint.Signature)
	if err != nil {
		return false, fmt.Errorf("failed to verify signature: %w", err)
	}

	return valid, nil
}

// EncodeCheckpoint encodes a checkpoint to signed note format
//
// Format:
//
//	<origin>
//	<tree-size>
//	<root-hash-base64>
//	<timestamp>
//
//	— <origin> <signature-base64>
func EncodeCheckpoint(checkpoint *Checkpoint) string {
	rootHashBase64 := base64.StdEncoding.EncodeToString(checkpoint.RootHash[:])
	signatureBase64 := base64.StdEncoding.EncodeToString(checkpoint.Signature)

	lines := []string{
		checkpoint.Origin,
		fmt.Sprintf("%d", checkpoint.TreeSize),
		rootHashBase64,
		fmt.Sprintf("%d", checkpoint.Timestamp),
		"",
		fmt.Sprintf("— %s %s", checkpoint.Origin, signatureBase64),
	}

	return strings.Join(lines, "\n")
}

// DecodeCheckpoint decodes a checkpoint from signed note format
func DecodeCheckpoint(encoded string) (*Checkpoint, error) {
	lines := strings.Split(strings.TrimSpace(encoded), "\n")

	if len(lines) < 6 {
		return nil, fmt.Errorf("invalid checkpoint format: too few lines (got %d, need at least 6)", len(lines))
	}

	origin := lines[0]

	var treeSize int64
	if _, err := fmt.Sscanf(lines[1], "%d", &treeSize); err != nil {
		return nil, fmt.Errorf("invalid tree size: %w", err)
	}

	rootHashBase64 := lines[2]

	var timestamp int64
	if _, err := fmt.Sscanf(lines[3], "%d", &timestamp); err != nil {
		return nil, fmt.Errorf("invalid timestamp: %w", err)
	}

	// Parse signature line: "— <origin> <signature>"
	signatureLine := lines[5]
	signatureRegex := regexp.MustCompile(`^— .+ (.+)$`)
	matches := signatureRegex.FindStringSubmatch(signatureLine)
	if len(matches) < 2 {
		return nil, fmt.Errorf("invalid checkpoint format: signature line malformed")
	}
	signatureBase64 := matches[1]

	// Decode base64
	rootHashBytes, err := base64.StdEncoding.DecodeString(rootHashBase64)
	if err != nil {
		return nil, fmt.Errorf("invalid root hash base64: %w", err)
	}

	if len(rootHashBytes) != HashSize {
		return nil, fmt.Errorf("invalid root hash length: %d (expected %d)", len(rootHashBytes), HashSize)
	}

	var rootHash [HashSize]byte
	copy(rootHash[:], rootHashBytes)

	signature, err := base64.StdEncoding.DecodeString(signatureBase64)
	if err != nil {
		return nil, fmt.Errorf("invalid signature base64: %w", err)
	}

	return &Checkpoint{
		TreeSize:  treeSize,
		RootHash:  rootHash,
		Timestamp: timestamp,
		Origin:    origin,
		Signature: signature,
	}, nil
}

// encodeCheckpointData encodes checkpoint data for signing
// Binary encoding: tree_size (8 bytes) + root_hash (32 bytes) + timestamp (8 bytes) + origin (variable)
func encodeCheckpointData(treeSize int64, rootHash [HashSize]byte, timestamp int64, origin string) []byte {
	originBytes := []byte(origin)
	buffer := make([]byte, 8+HashSize+8+len(originBytes))

	// Write tree size (64-bit big-endian)
	binary.BigEndian.PutUint64(buffer[0:8], uint64(treeSize))

	// Write root hash
	copy(buffer[8:8+HashSize], rootHash[:])

	// Write timestamp (64-bit big-endian)
	binary.BigEndian.PutUint64(buffer[8+HashSize:8+HashSize+8], uint64(timestamp))

	// Write origin
	copy(buffer[8+HashSize+8:], originBytes)

	return buffer
}

File: scitt-typescript/src/lib/merkle/checkpoints.ts
/**
 * Checkpoint (Signed Tree Head) Management
 *
 * A checkpoint is a signed commitment to the current state of the Merkle tree.
 * It contains:
 * - Tree size (number of entries)
 * - Root hash (Merkle tree root)
 * - Timestamp
 * - Origin (transparency service URL)
 * - Signature (signed by transparency service)
 *
 * Format follows the "signed note" format used by transparency logs.
 */

import { ES256Signer, ES256Verifier } from "../cose/signer.ts";

/**
 * Checkpoint structure
 */
export interface Checkpoint {
  treeSize: number;
  rootHash: Uint8Array;
  timestamp: number;
  origin: string;
  signature: Uint8Array;
}

/**
 * Create a checkpoint (signed tree head)
 *
 * @param treeSize - Number of entries in the tree
 * @param rootHash - Root hash of the Merkle tree (32 bytes)
 * @param privateKey - Private key for signing
 * @param origin - Transparency service origin URL
 * @returns Checkpoint with signature
 */
export async function createCheckpoint(
  treeSize: number,
  rootHash: Uint8Array,
  privateKey: CryptoKey,
  origin: string
): Promise<Checkpoint> {
  // Validate inputs
  if (rootHash.length !== 32) {
    throw new Error(`Root hash must be 32 bytes (SHA-256), got ${rootHash.length}`);
  }

  // Validate origin URL
  try {
    new URL(origin);
  } catch (error) {
    throw new Error(`Invalid origin URL: ${origin}`);
  }

  const timestamp = Date.now();

  // Create the data to be signed
  const dataToSign = encodeCheckpointData(treeSize, rootHash, timestamp, origin);

  // Sign with ES256
  const signer = new ES256Signer(privateKey);
  const signature = await signer.sign(dataToSign);

  return {
    treeSize,
    rootHash,
    timestamp,
    origin,
    signature,
  };
}

/**
 * Verify a checkpoint signature
 *
 * @param checkpoint - Checkpoint to verify
 * @param publicKey - Public key for verification
 * @returns True if signature is valid
 */
export async function verifyCheckpoint(
  checkpoint: Checkpoint,
  publicKey: CryptoKey
): Promise<boolean> {
  // Reconstruct the signed data
  const dataToSign = encodeCheckpointData(
    checkpoint.treeSize,
    checkpoint.rootHash,
    checkpoint.timestamp,
    checkpoint.origin
  );

  // Verify signature
  const verifier = new ES256Verifier(publicKey);
  return await verifier.verify(dataToSign, checkpoint.signature);
}

/**
 * Encode checkpoint to signed note format (text)
 *
 * Format (per RFC 6962 - root hash must be hex-encoded):
 * ```
 * <origin>
 * <tree-size>
 * <root-hash-hex>
 * <timestamp>
 *
 * — <origin> <signature-base64>
 * ```
 *
 * @param checkpoint - Checkpoint to encode
 * @returns Signed note string
 */
export function encodeCheckpoint(checkpoint: Checkpoint): string {
  // RFC 6962 requires hex encoding for merkle tree hashes
  const rootHashHex = Array.from(checkpoint.rootHash)
    .map(b => b.toString(16).padStart(2, '0'))
    .join('');
  const signatureBase64 = btoa(String.fromCharCode(...checkpoint.signature));

  const lines = [
    checkpoint.origin,
    checkpoint.treeSize.toString(),
    rootHashHex,
    checkpoint.timestamp.toString(),
    "",
    `— ${checkpoint.origin} ${signatureBase64}`,
  ];

  return lines.join("\n");
}

/**
 * Decode checkpoint from signed note format
 *
 * @param encoded - Signed note string
 * @returns Checkpoint
 */
export function decodeCheckpoint(encoded: string): Checkpoint {
  const lines = encoded.trim().split("\n");

  if (lines.length < 6) {
    throw new Error("Invalid checkpoint format: too few lines");
  }

  const origin = lines[0]!;
  const treeSize = parseInt(lines[1]!, 10);
  const rootHashStr = lines[2]!;
  const timestamp = parseInt(lines[3]!, 10);

  // Parse signature line: "— <origin> <signature>"
  const signatureLine = lines[5]!;
  const signatureMatch = signatureLine.match(/^— .+ (.+)$/);
  if (!signatureMatch) {
    throw new Error("Invalid checkpoint format: signature line malformed");
  }
  const signatureBase64 = signatureMatch[1]!;

  // Decode root hash - try hex first (RFC 6962 compliant), then base64 for backwards compatibility
  let rootHash: Uint8Array;

  // Check if it's hex (only contains 0-9, a-f, A-F)
  if (/^[0-9a-fA-F]+$/.test(rootHashStr) && rootHashStr.length === 64) {
    // Decode hex
    rootHash = new Uint8Array(32);
    for (let i = 0; i < 32; i++) {
      rootHash[i] = parseInt(rootHashStr.substr(i * 2, 2), 16);
    }
  } else {
    // Try base64 for backwards compatibility
    try {
      rootHash = Uint8Array.from(atob(rootHashStr), c => c.charCodeAt(0));
    } catch (error) {
      throw new Error(`Invalid root hash encoding (expected hex or base64): ${error}`);
    }
  }

  const signature = Uint8Array.from(atob(signatureBase64), c => c.charCodeAt(0));

  if (rootHash.length !== 32) {
    throw new Error(`Invalid root hash length: ${rootHash.length}`);
  }

  return {
    treeSize,
    rootHash,
    timestamp,
    origin,
    signature,
  };
}

/**
 * Encode checkpoint data for signing
 *
 * The signed data is a CBOR-encoded array:
 * [tree_size, root_hash, timestamp, origin]
 */
function encodeCheckpointData(
  treeSize: number,
  rootHash: Uint8Array,
  timestamp: number,
  origin: string
): Uint8Array {
  // Simple binary encoding: tree_size (8 bytes) + root_hash (32 bytes) + timestamp (8 bytes) + origin (variable)
  const encoder = new TextEncoder();
  const originBytes = encoder.encode(origin);

  const buffer = new Uint8Array(8 + 32 + 8 + originBytes.length);
  const view = new DataView(buffer.buffer);

  // Write tree size (64-bit big-endian)
  view.setBigUint64(0, BigInt(treeSize), false);

  // Write root hash
  buffer.set(rootHash, 8);

  // Write timestamp (64-bit big-endian)
  view.setBigUint64(40, BigInt(timestamp), false);

  // Write origin
  buffer.set(originBytes, 48);

  return buffer;
}
